File: 1071-greatest-common-divisor-of-strings/1071-greatest-common-divisor-of-strings.c
int gcd(int a, int b) {
    int tmp;
    while (b != 0) {
        tmp = a % b;
        a = b;
        b = tmp;
    }
    return a;
}

char * gcdOfStrings(char * str1, char * str2){
    size_t n;
    char *result;
    size_t s1_len = strlen(str1);
    size_t s2_len = strlen(str2);
    size_t concat_len = s1_len + s2_len + 1;
    char s1s2[concat_len];
    char s2s1[concat_len];
    
    memset(s1s2, '\0', concat_len);
    memset(s2s1, '\0', concat_len);
    
    strncpy(s1s2, str1, s1_len);
    strncat(s1s2, str2, s2_len);
    
    strncpy(s2s1, str2, s2_len);
    strncat(s2s1, str1, s1_len);
    
    if (strncmp(s1s2, s2s1, s1_len + s2_len) != 0)
        return calloc(sizeof(char), 1);
    
    n = gcd(s1_len, s2_len);
    result = malloc(sizeof(char) * n + 1);
    result[n] = '\0';
    strncpy(result, str1, n);
    return result;
}

File: 1071-greatest-common-divisor-of-strings/1071-greatest-common-divisor-of-strings.go
func gcdOfStrings(str1 string, str2 string) string {
    if str1 + str2 != str2 + str1 {
        return ""
    }
    gcd := calculateGCD(len(str1), len(str2))
    return str1[:gcd]
}

func calculateGCD(a, b int) int {
    for b != 0 {
        a, b = b, a % b
    }
    return a
}

/*
func gcdOfStrings(str1 string, str2 string) string {
    i := 0
    sub := ""
    maxSub := ""
    minLen := min(len(str1), len(str2))
    for i < minLen && str1[i] == str2[i] {
        sub += string(str1[i])
        if dividesString(str1, sub) && dividesString(str2, sub) {
            maxSub = sub
        }
        i++
    }
    
    return maxSub
}

func dividesString(s, sub string) bool {
    return strings.Repeat(sub, len(s)/len(sub)) == s
}

func min(a, b int) int {
    if a < b {
        return a
    }
    return b
}
*/